packages/messageport/tizen/src/messageport.h

```diff
@@ -2,59 +2,127 @@
-#include <flutter/standard_method_codec.h>
 #include <message_port.h>
 
+#include <functional>
 #include <map>
-#include <set>
+#include <memory>
+#include <optional>
+#include <string>
+#include <variant>
 
-typedef std::unique_ptr<flutter::EventSink<flutter::EncodableValue>> EventSink;
+class MessagePortError {
+ public:
+  explicit MessagePortError(int error_code)
+      : error_code_(error_code),
+        error_message_(get_error_message(error_code_)) {}
+
+  const std::string& message() const { return error_message_; }
+
+  int code() { return error_code_; }
+
+ private:
+  int error_code_;
+  std::string error_message_;
+};
+
+template <class T>
+class ErrorOr {
+ public:
+  ErrorOr(const T& other) { new (&vlaue_or_error_) T(other); }
 
-struct MessagePortResult {
-  MessagePortResult() : error_code(MESSAGE_PORT_ERROR_NONE){};
-  MessagePortResult(int code) : error_code(code) {}
+  ErrorOr(const T&& other) { vlaue_or_error_ = std::move(other); }
 
-  // Returns false on error
-  operator bool() const { return MESSAGE_PORT_ERROR_NONE == error_code; }
+  ErrorOr(const MessagePortError& other) {
+    new (&vlaue_or_error_) MessagePortError(other);
+  }
 
-  std::string message() { return get_error_message(error_code); }
+  ErrorOr(const MessagePortError&& other) {
+    vlaue_or_error_ = std::move(other);
+  }
+
+  bool has_error() const {
+    return std::holds_alternative<MessagePortError>(vlaue_or_error_);
+  }
+
+  const T& value() const { return std::get<T>(vlaue_or_error_); };
+
+  const MessagePortError& error() const {
+    return std::get<MessagePortError>(vlaue_or_error_);
+  };
+
+ private:
+  ErrorOr() = default;
+
+  std::variant<T, MessagePortError> vlaue_or_error_;
+};
 
-  int error_code;
+struct Message {
+  const std::string remote_app_id;
+  const std::string remote_port;
+  bool trusted_remote_port = false;
+  std::unique_ptr<std::vector<uint8_t>> encoded_message;
+  const std::string error;
 };
 
-class MessagePortManager {
+typedef std::function<void(const Message&)> OnMessage;
+
+class LocalPort {
  public:
-  MessagePortManager();
-  ~MessagePortManager();
-
-  MessagePortResult CheckRemotePort(std::string& remote_app_id,
-                                    std::string& port_name, bool is_trusted,
-                                    bool* result);
-  MessagePortResult RegisterLocalPort(const std::string& port_name,
-                                      EventSink sink, bool is_trusted,
-                                      int* local_port);
-  MessagePortResult UnregisterLocalPort(int local_port_id);
-  MessagePortResult Send(std::string& remote_app_id, std::string& port_name,
-                         flutter::EncodableValue& message, bool is_trusted);
-  MessagePortResult Send(std::string& remote_app_id, std::string& port_name,
-                         flutter::EncodableValue& message, bool is_trusted,
-                         int local_port);
+  LocalPort(const std::string name, const bool is_trusted)
```

Comment:
```suggestion
  LocalPort(const std::string& name, bool is_trusted)
```
The second `const` doesn't seem useful.

The same applies to `RemotePort::RemotePort`.